### PR TITLE
Make ARN builders require account ID and region args

### PR DESCRIPTION
**Pull request by viren-nadkarni**
## Motivation

ARNs in LocalStack were incorrect in certain situations because they used account ID and region from the thread context via the use of `get_aws_account_id()` and `aws_stack.get_region()`.

## Implementation

This PR is a complete refactoring of our ARN builder utilities that make the `account_id` and `region_name` field mandatory. The correct values for these are passed down the from the request context.

## Related

Depends on:
- https://github.com/localstack/localstack/pull/9400
- https://github.com/localstack/localstack/pull/9401
- https://github.com/localstack/localstack/pull/9418

### CODE REVIEW

File: localstack/services/apigateway/models.py
Comment (viren-nadkarni) [inline localstack/services/apigateway/models.py:88]: This mimics the current behaviour where the account ID of the first APIGW request when the provider starts up. The effort involved to refactor APIGW codebase to fix this is substantial and out of the scope of this PR and will be addressed in the future.

File: localstack/services/cloudformation/models/cloudformation.py
Comment (viren-nadkarni) [inline localstack/services/cloudformation/models/cloudformation.py:154]: The value for stack ID `id-123` is picked up from the existing default argument.

File: localstack/services/ecr/resource_providers/aws_ecr_repository.py
Comment (viren-nadkarni) [inline localstack/services/ecr/resource_providers/aws_ecr_repository.py:102]: This is a mock repo so it's okay to use the default values.

File: localstack/services/lambda_/lambda_api.py
Comment (viren-nadkarni) [inline localstack/services/lambda_/lambda_api.py:2341]: For the legacy Lambda provider, again the effort involved to ensure ARNs are correct is substantial hence we resort to default values.

File: localstack/services/lambda_/lambda_api.py
Comment (viren-nadkarni) [inline localstack/services/lambda_/lambda_api.py:2342]: For the legacy Lambda provider, the effort involved to refactor the codebase is substantial so we make an exception and use the `get_region()` and `get_aws_account_id()` helpers.

File: tests/aws/services/stepfunctions/legacy/test_stepfunctions_legacy.py
Comment (viren-nadkarni) [inline tests/aws/services/stepfunctions/legacy/test_stepfunctions_legacy.py:306]: The third argument of this function was actually called `env=` and was unused within the function.

This is now removed.

File: localstack/services/apigateway/models.py
Comment (bentsku) [reply localstack/services/apigateway/models.py:88]: I see the problem here, the `_account_id` field of the store is not passed to the `__init__` called but manually attached later on to the store, so we can't access it here. Was there a specific reason it was done this way?

I don't think this mimics the current behaviour though:
```bash
$ awslocal apigateway get-account
{
    "cloudwatchRoleArn": "arn:aws:iam::000000000000:role/api-gw-cw-role",
    ...
}
$ AWS_ACCESS_KEY_ID=000000000002 awslocal apigateway get-account
{
    "cloudwatchRoleArn": "arn:aws:iam::000000000002:role/api-gw-cw-role",
    ...
}
```
This might create some issues, but I don't think we currently have the CloudWatch integration with APIGW, so it shouldn't be an IAM issue I hope.

Do you see how this could be achieved with the current architecture? Even a big a refactor in APIGW might not solve this if `ApiGatewayStore.account` is called in different places, we would need to manually update the account after the store instantiation. How can we address this issue?

File: localstack/services/dynamodb/provider.py
Comment (bentsku) [inline localstack/services/dynamodb/provider.py:1759]: 🎉

File: tests/aws/services/apigateway/test_apigateway_basic.py
Comment (bentsku) [inline tests/aws/services/apigateway/test_apigateway_basic.py:1983]: nit: I think we could use the fixture here for the sqs queue ARN.

**Review by bentsku — COMMENTED** [review]
Wow, quite an extensive set of changes, that's awesome! This is really a great!
I just have a question concerning the ApiGatewayStore instantiation with default values. Will we have a solution in the future for this use case? I'd like to know if we can have an idea of what to do to fix this issue before changing it that way.

File: localstack/services/apigateway/models.py
Comment (viren-nadkarni) [reply localstack/services/apigateway/models.py:88]: > Was there a specific reason it was done this way?

IIRC this was done to avoid large scale code refactoring, you see `RegionBackend`s, the precursor to `Store`, defined `__init__` functions. We wanted to introduce stores in a backward compatible manner.

> Do you see how this could be achieved with the current architecture?

Basically this ties back to the previous question, the most straightforward way is not to define business logic in the store definition. Initialiser functions and other member methods with business logic must not be defined as part of the store definition.

File: tests/aws/services/apigateway/test_apigateway_basic.py
Comment (viren-nadkarni) [reply tests/aws/services/apigateway/test_apigateway_basic.py]: This is resolved on the merge from `master`

File: localstack/services/cloudformation/models/cloudformation.py
Comment (steffyP) [reply localstack/services/cloudformation/models/cloudformation.py:154]: sorry, just realized that the definition changed as well :) but maybe it makes sense then to use a short-uid here instead? I am not sure why it was hard-coded in the first place though 🤔 Do you have any background information maybe?

**Review by steffyP — APPROVED** [review]
Thanks for cleaning up the usage of region and account-ids 🚀 👍 
The changes LGTM.

Just added a question regarding the usage of the default stack ID `id-123` for the cloudformation stack arn.

File: localstack/services/cloudformation/models/cloudformation.py
Comment (viren-nadkarni) [reply localstack/services/cloudformation/models/cloudformation.py:154]: > Do you have any background information maybe?

I don't know I'm afraid! I was hesitant to change it to `short_uid()` because I grepped it and saw it being used in some snapshots.

Comment (steffyP) [reply localstack/services/cloudformation/models/cloudformation.py:154]: I see. Maybe we can consult with @dominikschubert when he is back. 
Anyhow, this should not be blocking the PR.

File: localstack/services/apigateway/models.py
Comment (bentsku) [reply localstack/services/apigateway/models.py:88]: I understand. I suppose we could get a quick fix once this is merged in `get_account` or something similar, to initialize the account value in the store if it doesn't exist instead. Maybe we could get something quick that wouldn't need a big refactor, I'll try to make something once this is merged. Thanks for the precisions!

**Review by bentsku — APPROVED** [review]
LGTM! I'll try to get something up to fix the APIGateway issue in the coming days. Thanks again for tackling this!

**Pull request merged**
